src/cmdlib.sh

```diff
@@ -318,8 +321,20 @@ openshift_git_hack() {
     info "NOTICE: using workaround for Openshift missing .git"
 
     if [ "${GIT_URL:-x}" == "x" ] ||  [ "${GIT_REF:-x}" ]; then
-        info "Checking out bare git of ${GIT_URL} to ${gitd}/.git, ref ${GIT_REF}"
-        git clone --depth 1 --bare -b "${GIT_REF}" "${GIT_URL}" "${gitd}/.git"
+        tmpgit="$(mktemp -d)"
+        trap 'rm -rf ${tmpgit}; unset tmpgit;' EXIT
+
+        info "Re-creating git tree in ${gitd} from ${GIT_URL}, ref ${GIT_REF}"
+        git clone --depth 1 -b "${GIT_REF}" "${GIT_URL}" "${tmpgit}"/clone
+
+        if test -f "${gitd}"/.gitmodules; then
+            pushd "${tmpgit}"/clone
+            info "Ensuring git modules exists too"
+            git submodule update --init --force
+            popd
+        fi
+
+        rsync -av --ignore-existing "${tmpgit}"/clone/ "${gitd}"/
```

Comment:
For the record, I strongly dislike this approach, but its the best solution and ensures that only missing files are added back -- but that could mean that someone who triggers this code path (need to have envVars `OPENSHIFT_GIT_HACK`, `GIT_REF` and `GIT_URL` specifically defined) could have deleted files added back. But this is the only real option with EL7.  The only reason I think this is an acceptable solution is because the envVars should only be used when executing in a CI situation, and the source that is being mangled is clean already.

Reply:
Understood

Reply:
can we add a comment above this rsync with some of that context?

Reply:
@dustymabe I re-worked the the comments, rebased and pushed.